### DISCUSSION TITLE
Help with MCP23s17 port expanders

### DISCUSSION THREAD

**Discussion by raspitoaster** [arendst/Tasmota#21393] [Q&A] [answered]
Hi - I know that I2C port expanders (MCP23008 and MCP23017) are supported for quite a long time.
Since v12.4.0.2, there is obviously some kind of support for the SPI versions (MCP23s08 and MCP23s17).
According to https://tasmota.github.io/docs/MCP230xx/#mode-2 you just have to add  #define USE_MCP23XXX_DRV to the user_config_override.h

I use the great tasmocompiler to build my own firmwares - and I tried to include the MCP23s17 driver with Tasmota 13.4.0.

When only adding
#define USE_MCP23XXX_DRV,
the build stops with an error like "base address of MCP230xx not defined".

When adding
#define USE_MCP23XXX_DRV
#define USE_MCP230xx_ADDR 0x00
the build finishes and I get a working firmware.

Using this firmware on a NodeMCU Amica v2 (8266 4M) with SPI configured as SSPI, there is a console message after boot:

00:00:00.093 SPI: Software using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)

"CS" is configured on GPIO15 and I also tried with "MCP23Sxx CS" "1" on GPIO15, but both without any success.
There is no console message "MCP23xx detected on xyz" like with the I2C expanders.

Like mentioned in the docs, Tasmota complained about missing Mode-2 config, so I added a file mcp23x.dat
{"NAME":"MCP23S17 Expander","BASE":0,"GPIO":[32,33,34,35,36,37,38,39,224,225,226,227,228,229,230,231]}

No more complains, but no success though. Do I miss something?
In the docs, there is an example for Mode-2 mentioning Shelly 4PM, so I assume that MCP23s17 is somehow supported
{"NAME":"MCP23S17 Shelly Pro 4PM","GPIO":[194,193,65,66,3840,64,192,0,224,0,0,0,227,225,226,195]}

Thanks

**Comment by HofeBY2** [comment]
What's the output of command I2Cdriver ?

**Reply by raspitoaster** [reply]
13:04:53.553 CMD: I2Cdriver
13:04:53.562 MQT: nodemcu-005/stat/RESULT = {"I2CDriver":"8,9,10,15,20,21,22,32,36,41,42,43,45,51,52,60,77"}

**Comment by arendst** [comment]
Try hardware SPI (``SPI``) instead of software SPI (``SSPI``)

**Reply by raspitoaster** [reply]
Ah, I thought SSPI is "second SPI" which is usable and SPI is connected to the flash chip, but "software SPI" makes sense.
Assigning the same pins to SPI instead of SSPI gives the output
00:00:00.098 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)

What is the purpose of this "MCP23Sxx CS" with number "1" to "9" in the module configuration page?
When I substitute "SPI CS" with "MCP23Sxx CS" on GPIO15, the above message is no more displayed.

FYI: I connected A0,A1,A2 to GND and RESET to 3V3 - but there is no difference to if I do not connect the pins.

**Reply by sfromis** [reply]
The purpose of "MCP23Sxx CS" is to let Tasmota know that you have a MCP23Sxx chip on the SPI bus activated by that chip select pin. The number is when having multiple chips. The generic "SPI CS" does not indicate what the hardware is.

**Reply by raspitoaster** [reply]
So you advise to use "MCP23Sxx CS"? 
But the result seems to be even worse - with "SPI CS" there is a line of output, with "MCP23Sxx CS" there is nothing.
I am also not sure about the address to use during compile time (in user_config_override.h) because the 0x20 mentioned everywhere in the docs is the base address of the I2C MCP230xx ... I am using the chip with address 0x00.
Maybe this is the problem? With "MCP23Sxx CS" I can only choose between "1" and "9" - but not "0".
Do I have to compile the driver e.g. with 0x01 and set the A1,A2 to GND and A2 to 3V3?

**Reply by sfromis** [reply]
Whatever you do, that `0x00` does not make sense, even for the preferable I2C. I'd have used `0x20` (the default), but if your board is really for SPI (and this is supported also on ESP8266), the I2C address has no meaning. And the "correct" CS number to use is 1-9 in sequence, starting with 1 for first chip. What you do with those pins on the board IDK. Reading data sheet is generally a good plan...

**Comment by arendst** [comment]
I need to chk if MCP23s is supported on esp8266. I prefer I2C ofcourse.

**Reply by arendst** [reply]
Yeah. It won't work with MCP23S on ESP8266.

I'll remove the test for _CS on ESP8266 too like I did long time agon on ESP32. Hold on and try latest dev within 10 minutes...

EDIT: Wrong assumption. It should work.

**Reply by arendst** [reply, marked as answer]
Try latest dev which removed the SPI CS test.

Keep A0,A1 and A2 to Gnd and use ``MCP23Sxx CS 1``

**Reply by arendst** [reply]
BTW the "old" dev release also works fine:

![image](https://github.com/arendst/Tasmota/assets/11044339/3f5a205f-3c33-4453-b066-84717bc01f2f)

```
00:00:00.001 HDW: ESP8266EX
00:00:00.019 UFS: FlashFS mounted with 1984 kB free
00:00:00.068-028 CFG: Loaded from File, Count 72
00:00:00.073-028 QPC: Count 1
00:00:00.078-028 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.085-028 MCP: Invalid template
00:00:00.087-028 ROT: Mode 1
00:00:00.089-028 SRC: Restart
00:00:00.091-028 Project tasmota - Wemos11 Version 13.4.1.2(theo)-2_7_6(2024-05-09T14:07:34)
00:00:00.183-028 WIF: Attempting connection...
00:00:00.186-026 WIF: Network (re)scan started...
00:00:01.107-027 WIF: Attempting connection...
00:00:02.106-027 WIF: Attempting connection...
00:00:03.106-027 WIF: Attempting connection...
00:00:03.107-026 WIF: Network 0, AP1, SSId indebuurt_IoT, Channel 11, BSSId 18:E8:29:CA:17:C1, RSSI -29, Encryption 1
00:00:04.001-028 WIF: Connecting to AP1 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as wemos11...
00:00:05.482-027 WIF: Connected
00:00:05.734-024 HTP: Web server active on wemos11 with IP address 192.168.2.214
00:00:06.349-025 QPC: Reset
00:00:06.445-024 RTC: UTC 2024-05-09T12:10:51Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
13:10:51.000-025 RTC: Synced by NTP
13:10:51.287-025 MQT: Attempting connection...
13:10:51.310-024 MQT: Connected
13:10:51.313-021 MQT: tele/wemos11/LWT = Online (retained)
13:10:51.315-021 MQT: cmnd/wemos11/POWER = 
13:10:51.317-021 MQT: Subscribe to cmnd/wemos11/#
13:10:51.318-021 MQT: Subscribe to cmnd/sonoffs/#
13:10:51.321-021 MQT: Subscribe to cmnd/DVES_17DA05_fb/#
13:10:51.328-021 MQT: tele/wemos11/INFO1 = {"Info1":{"Module":"Generic","Version":"13.4.1.2(theo)","FallbackTopic":"cmnd/DVES_17DA05_fb/","GroupTopic":"cmnd/sonoffs/"}}
13:10:51.343-021 MQT: tele/wemos11/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"wemos11","IPAddress":"192.168.2.214"}}
13:10:51.353-021 MQT: tele/wemos11/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":39}}
13:10:53.297-024 APP: Boot Count 39
```

So I'll revert latest change.

**Reply by raspitoaster** [reply]
Now I tried dev (Tasmota Version 13.4.1.2(TasmoCompiler-esp82664M)-2_7_6(2024-05-10T07:34:28) : 

00:00:00.319 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)

09:39:55.318 MQT: nodemcu-005/stat/RESULT = {"I2CDriver":"22,77"}

Still no sign of life from the MCP23s17.

Then instead of "SPI CS", I used "MCP23Sxx CS" on GPIO15 : 
**--> now there are 9 relays and 9 buttons displayed**
1- is the button/relay_i (LED) on the nodemcu
2-9 - are the buttons/relais on the mcp

mcp23x.dat :
{"NAME":"MCP23S17 Expander","BASE":0,"GPIO":[32,33,34,35,36,37,38,39,224,225,226,227,228,229,230,231]}

I can toggle the PINs defined as outputs (GPB0-7) from the web page and from the console.
I can use the PINs defined as buttons (GPA0-7) to toggle the relais.

100% success - thank you very much!
